### PR TITLE
[DSIP-19] Support k8s connections in the connection center, as well as external connections to the connection center in k8s tasks

### PR DESCRIPTION
**Pull request by xdu-chenrj**
issue: https://github.com/apache/dolphinscheduler/issues/14338

Related issues: [DSIP-19][Feature] Add connection center feature for DS https://github.com/apache/dolphinscheduler/issues/10283

Mail: https://lists.apache.org/thread/xl6pb3sbrt0ffrf1fltcph39s9w1pjlx

### CODE REVIEW

**Comment by EricGao888** [comment]
@xdu-chenrj The commit history seems not correct.

**Comment by xdu-chenrj** [comment]
> Copy link
> Quote reply

yes

**Comment by xdu-chenrj** [comment]
> @xdu-chenrj The commit history seems not correct.

yes

**Review by EricGao888 — DISMISSED** [review]
Waiting for CI to pass

**Comment by EricGao888** [comment]
@xdu-chenrj 
image
Commit history seems still incorrect.

**Review by EricGao888 — APPROVED** [review]

File: dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/am/KubernetesApplicationManager.java [hunk 1]
```diff
@@ -131,6 +132,8 @@ private FilterWatchListDeletable<Pod, PodList, PodResource> getListenPod(Kuberne
     private KubernetesClient getClient(KubernetesApplicationManagerContext kubernetesApplicationManagerContext) {
         K8sTaskExecutionContext k8sTaskExecutionContext =
                 kubernetesApplicationManagerContext.getK8sTaskExecutionContext();
+        k8sTaskExecutionContext
+                .setConfigYaml(JSONUtils.getNodeString(k8sTaskExecutionContext.getConnectionParams(), "kubeConfig"));
```
Comment (Radeity) [inline dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/am/KubernetesApplicationManager.java:136]: You can not directly change this. For normal task, like spark on k8s, `K8sTaskExecutionContext` is constructed by `TaskExecutionContextFactory`, when they call `getClient`, `connectionParams` is null.

File: dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/parameters/K8sTaskParameters.java [hunk 2]
```diff
@@ -44,14 +51,29 @@ public class K8sTaskParameters extends AbstractParameters {
...
+        k8sTaskExecutionContext.setConnectionParams(
+                Objects.nonNull(dataSourceParameters) ? dataSourceParameters.getConnectionParams() : null);
+        return k8sTaskExecutionContext;
     }
-
     @Override
     public List<ResourceInfo> getResourceFilesList() {
         return new ArrayList<>();
     }
+    @Override
```
Comment (Radeity) [inline dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/parameters/K8sTaskParameters.java:73]: Better leave a blank line before each method :D

**Comment by Radeity** [comment]
Hi @xdu-chenrj,  non-K8S task should better be refactored and use k8s connection together with k8s task, maybe we have to revert this PR, or to fix this bug (refactor) ASAP.  cc @EricGao888

**Comment by xdu-chenrj** [comment]
> Hi @xdu-chenrj, non-K8S task should better be refactored and use k8s connection together with k8s task, maybe we have to revert this PR, or to fix this bug (refactor) ASAP. cc @EricGao888

ok

**Comment by EricGao888** [comment]
> Hi @xdu-chenrj, non-K8S task should better be refactored and use k8s connection together with k8s task, maybe we have to revert this PR, or to fix this bug (refactor) ASAP. cc @EricGao888

@xdu-chenrj I suggest submitting a patch to fix it instead of reverting the whole PR. BTW, as the PR passed the CI, it seems the UT hadn't covered this case. I think it will be better to add related UT to cover this scenario for those tasks using k8s connection.

File: dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/am/KubernetesApplicationManager.java (hunk 1 above)
Comment (xdu-chenrj) [reply dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/am/KubernetesApplicationManager.java:136]: Thank you for your suggestion

**Comment by xdu-chenrj** [comment]
> > Hi @xdu-chenrj, non-K8S task should better be refactored and use k8s connection together with k8s task, maybe we have to revert this PR, or to fix this bug (refactor) ASAP. cc @EricGao888
> 
> @xdu-chenrj I suggest submitting a patch to fix it instead of reverting the whole PR. BTW, as the PR passed the CI, it seems the UT hadn't covered this case. I think it will be better to add related UT to cover this scenario for those tasks using k8s connection.

Yes, thank you for your suggestion

**Comment by Radeity** [comment]
> > > Hi @xdu-chenrj, non-K8S task should better be refactored and use k8s connection together with k8s task, maybe we have to revert this PR, or to fix this bug (refactor) ASAP. cc @EricGao888
> > 
> > 
> > @xdu-chenrj I suggest submitting a patch to fix it instead of reverting the whole PR. BTW, as the PR passed the CI, it seems the UT hadn't covered this case. I think it will be better to add related UT to cover this scenario for those tasks using k8s connection.
> 
> Yes, thank you for your suggestion

You can see `TaskExecutionContextFactory.setK8sTaskRelation`, there are several types of task which already construct `K8sTaskExecutionContext` in `Master`. Anyway, you can ping me anytime if you need help with this patch :D

**Comment by EricGao888** [comment]
> > > > Hi @xdu-chenrj, non-K8S task should better be refactored and use k8s connection together with k8s task, maybe we have to revert this PR, or to fix this bug (refactor) ASAP. cc @EricGao888
> > > 
> > > 
> > > @xdu-chenrj I suggest submitting a patch to fix it instead of reverting the whole PR. BTW, as the PR passed the CI, it seems the UT hadn't covered this case. I think it will be better to add related UT to cover this scenario for those tasks using k8s connection.
> > 
> > 
> > Yes, thank you for your suggestion
> 
> You can see `TaskExecutionContextFactory.setK8sTaskRelation`, there are several types of task which already construct `K8sTaskExecutionContext` in `Master`. Anyway, you can ping me anytime if you need help with this patch :D

pretty cool

**Comment by xdu-chenrj** [comment]
> You can see `TaskExecutionContextFactory.setK8sTaskRelation`, there are several types of task which already construct `K8sTaskExecutionContext` in `Master`. Anyway, you can ping me anytime if you need help with this patch :D

Thank you very much for your suggestion  :D

**Comment by xdu-chenrj** [comment]
> Quote reply

```java
    private K8sTaskExecutionContext setK8sTaskRelation(TaskInstance taskInstance) {
        K8sTaskExecutionContext k8sTaskExecutionContext = null;
        String namespace = "", kubeConfig = "";
        log.info("getTaskParams {}", taskInstance.getTaskParams());
        switch (taskInstance.getTaskType()) {
            case "K8S":
            case "KUBEFLOW":
                K8sTaskParameters k8sTaskParameters =
                        JSONUtils.parseObject(taskInstance.getTaskParams(), K8sTaskParameters.class);
                namespace = k8sTaskParameters.getNamespace();
                kubeConfig = k8sTaskParameters.getKubeConfig();
                break;
            case "SPARK":
                SparkParameters sparkParameters =
                        JSONUtils.parseObject(taskInstance.getTaskParams(), SparkParameters.class);
                if (StringUtils.isNotEmpty(sparkParameters.getNamespace())) {
                    namespace = sparkParameters.getNamespace();
                }
                break;
            default:
                break;
        }

        if (StringUtils.isNotEmpty(namespace)) {
            String configYaml = kubeConfig;
            log.info("String configYaml {}", configYaml);
            if (configYaml != null) {
                k8sTaskExecutionContext =
                        new K8sTaskExecutionContext(configYaml, JSONUtils.toMap(namespace).get(NAMESPACE_NAME));
            }
        }
        return k8sTaskExecutionContext;
    } 
```
hi, I want to assign the value `configYaml` to `k8sTaskExecutionContext` through this method, but `taskInstance. getTaskParams()` lacks the `kubeConfig` required in `K8sTaskParameters`. I would like to receive your specific advice

**Comment by xdu-chenrj** [comment]
> > > > Hi @xdu-chenrj, non-K8S task should better be refactored and use k8s connection together with k8s task, maybe we have to revert this PR, or to fix this bug (refactor) ASAP. cc @EricGao888
> > > 
> > > 
> > > @xdu-chenrj I suggest submitting a patch to fix it instead of reverting the whole PR. BTW, as the PR passed the CI, it seems the UT hadn't covered this case. I think it will be better to add related UT to cover this scenario for those tasks using k8s connection.
> > 
> > 
> > Yes, thank you for your suggestion
> 
> You can see `TaskExecutionContextFactory.setK8sTaskRelation`, there are several types of task which already construct `K8sTaskExecutionContext` in `Master`. Anyway, you can ping me anytime if you need help with this patch :D


```java
    private K8sTaskExecutionContext setK8sTaskRelation(TaskInstance taskInstance) {
        K8sTaskExecutionContext k8sTaskExecutionContext = null;
        String namespace = "", kubeConfig = "";
        log.info("getTaskParams {}", taskInstance.getTaskParams());
        switch (taskInstance.getTaskType()) {
            case "K8S":
            case "KUBEFLOW":
                K8sTaskParameters k8sTaskParameters =
                        JSONUtils.parseObject(taskInstance.getTaskParams(), K8sTaskParameters.class);
                namespace = k8sTaskParameters.getNamespace();
                kubeConfig = k8sTaskParameters.getKubeConfig();
                break;
            case "SPARK":
                SparkParameters sparkParameters =
                        JSONUtils.parseObject(taskInstance.getTaskParams(), SparkParameters.class);
                if (StringUtils.isNotEmpty(sparkParameters.getNamespace())) {
                    namespace = sparkParameters.getNamespace();
                }
                break;
            default:
                break;
        }

        if (StringUtils.isNotEmpty(namespace)) {
            String configYaml = kubeConfig;
            log.info("String configYaml {}", configYaml);
            if (configYaml != null) {
                k8sTaskExecutionContext =
                        new K8sTaskExecutionContext(configYaml, JSONUtils.toMap(namespace).get(NAMESPACE_NAME));
            }
        }
        return k8sTaskExecutionContext;
    } 
```
hi, I want to assign the value `configYaml` to `k8sTaskExecutionContext` through this method, but `taskInstance. getTaskParams()` lacks the `kubeConfig` required in `K8sTaskParameters`. I would like to receive your specific advice

**Comment by Radeity** [comment]
> 
> ```java
>     private K8sTaskExecutionContext setK8sTaskRelation(TaskInstance taskInstance) {
>         K8sTaskExecutionContext k8sTaskExecutionContext = null;
>         String namespace = "", kubeConfig = "";
>         log.info("getTaskParams {}", taskInstance.getTaskParams());
>         switch (taskInstance.getTaskType()) {
>             case "K8S":
>             case "KUBEFLOW":
>                 K8sTaskParameters k8sTaskParameters =
>                         JSONUtils.parseObject(taskInstance.getTaskParams(), K8sTaskParameters.class);
>                 namespace = k8sTaskParameters.getNamespace();
>                 kubeConfig = k8sTaskParameters.getKubeConfig();
>                 break;
>             case "SPARK":
>                 SparkParameters sparkParameters =
>                         JSONUtils.parseObject(taskInstance.getTaskParams(), SparkParameters.class);
>                 if (StringUtils.isNotEmpty(sparkParameters.getNamespace())) {
>                     namespace = sparkParameters.getNamespace();
>                 }
>                 break;
>             default:
>                 break;
>         }
> 
>         if (StringUtils.isNotEmpty(namespace)) {
>             String configYaml = kubeConfig;
>             log.info("String configYaml {}", configYaml);
>             if (configYaml != null) {
>                 k8sTaskExecutionContext =
>                         new K8sTaskExecutionContext(configYaml, JSONUtils.toMap(namespace).get(NAMESPACE_NAME));
>             }
>         }
>         return k8sTaskExecutionContext;
>     } 
> ```
> 
> hi, I want to assign the value `configYaml` to `k8sTaskExecutionContext` through this method, but `taskInstance. getTaskParams()` lacks the `kubeConfig` required in `K8sTaskParameters`. I would like to receive your specific advice

Can we move method `K8sTaskParameters .generateExtendedContext` to `AbstractParameters` and name it `generateK8sTaskExecutionContext`. Then, for KUBEFLOW, Kubernetes, Spark task, call `generateK8sTaskExecutionContext` to construct `K8sTaskExecutionContext` inside their parameters' construction function. In this way, we don't have to `setK8sTaskRelation` in `Master`, WDYT?

**Comment by xdu-chenrj** [comment]
> > ```java
> >     private K8sTaskExecutionContext setK8sTaskRelation(TaskInstance taskInstance) {
> >         K8sTaskExecutionContext k8sTaskExecutionContext = null;
> >         String namespace = "", kubeConfig = "";
> >         log.info("getTaskParams {}", taskInstance.getTaskParams());
> >         switch (taskInstance.getTaskType()) {
> >             case "K8S":
> >             case "KUBEFLOW":
> >                 K8sTaskParameters k8sTaskParameters =
> >                         JSONUtils.parseObject(taskInstance.getTaskParams(), K8sTaskParameters.class);
> >                 namespace = k8sTaskParameters.getNamespace();
> >                 kubeConfig = k8sTaskParameters.getKubeConfig();
> >                 break;
> >             case "SPARK":
> >                 SparkParameters sparkParameters =
> >                         JSONUtils.parseObject(taskInstance.getTaskParams(), SparkParameters.class);
> >                 if (StringUtils.isNotEmpty(sparkParameters.getNamespace())) {
> >                     namespace = sparkParameters.getNamespace();
> >                 }
> >                 break;
> >             default:
> >                 break;
> >         }
> > 
> >         if (StringUtils.isNotEmpty(namespace)) {
> >             String configYaml = kubeConfig;
> >             log.info("String configYaml {}", configYaml);
> >             if (configYaml != null) {
> >                 k8sTaskExecutionContext =
> >                         new K8sTaskExecutionContext(configYaml, JSONUtils.toMap(namespace).get(NAMESPACE_NAME));
> >             }
> >         }
> >         return k8sTaskExecutionContext;
> >     } 
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > hi, I want to assign the value `configYaml` to `k8sTaskExecutionContext` through this method, but `taskInstance. getTaskParams()` lacks the `kubeConfig` required in `K8sTaskParameters`. I would like to receive your specific advice
> 
> Can we move method `K8sTaskParameters .generateExtendedContext` to `AbstractParameters` and name it `generateK8sTaskExecutionContext`. Then, for KUBEFLOW, Kubernetes, Spark task, call `generateK8sTaskExecutionContext` to construct `K8sTaskExecutionContext` inside their parameters' construction function. In this way, we don't have to `setK8sTaskRelation` in `Master`, WDYT?

I fixed this issue, and then I will create an issue and PR

**Comment by xdu-chenrj** [comment]
> > > ```java
> > >     private K8sTaskExecutionContext setK8sTaskRelation(TaskInstance taskInstance) {
> > >         K8sTaskExecutionContext k8sTaskExecutionContext = null;
> > >         String namespace = "", kubeConfig = "";
> > >         log.info("getTaskParams {}", taskInstance.getTaskParams());
> > >         switch (taskInstance.getTaskType()) {
> > >             case "K8S":
> > >             case "KUBEFLOW":
> > >                 K8sTaskParameters k8sTaskParameters =
> > >                         JSONUtils.parseObject(taskInstance.getTaskParams(), K8sTaskParameters.class);
> > >                 namespace = k8sTaskParameters.getNamespace();
> > >                 kubeConfig = k8sTaskParameters.getKubeConfig();
> > >                 break;
> > >             case "SPARK":
> > >                 SparkParameters sparkParameters =
> > >                         JSONUtils.parseObject(taskInstance.getTaskParams(), SparkParameters.class);
> > >                 if (StringUtils.isNotEmpty(sparkParameters.getNamespace())) {
> > >                     namespace = sparkParameters.getNamespace();
> > >                 }
> > >                 break;
> > >             default:
> > >                 break;
> > >         }
> > > 
> > >         if (StringUtils.isNotEmpty(namespace)) {
> > >             String configYaml = kubeConfig;
> > >             log.info("String configYaml {}", configYaml);
> > >             if (configYaml != null) {
> > >                 k8sTaskExecutionContext =
> > >                         new K8sTaskExecutionContext(configYaml, JSONUtils.toMap(namespace).get(NAMESPACE_NAME));
> > >             }
> > >         }
> > >         return k8sTaskExecutionContext;
> > >     } 
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > hi, I want to assign the value `configYaml` to `k8sTaskExecutionContext` through this method, but `taskInstance. getTaskParams()` lacks the `kubeConfig` required in `K8sTaskParameters`. I would like to receive your specific advice
> > 
> > 
> > Can we move method `K8sTaskParameters .generateExtendedContext` to `AbstractParameters` and name it `generateK8sTaskExecutionContext`. Then, for KUBEFLOW, Kubernetes, Spark task, call `generateK8sTaskExecutionContext` to construct `K8sTaskExecutionContext` inside their parameters' construction function. In this way, we don't have to `setK8sTaskRelation` in `Master`, WDYT?
> 
> I fixed this issue, and then I will create an issue and PR

and I plan to submit a PR to fix the problem first, and then add UT. Is that okay?

**Comment by Radeity** [comment]
> I fixed this issue, and then I will create an issue and PR

Glad to hear that!

> and I plan to submit a PR to fix the problem first, and then add UT. Is that okay?

Sure, you can. I think, for a single pr, sometimes, small is better :D

**Comment by xdu-chenrj** [comment]
> > I fixed this issue, and then I will create an issue and PR
> 
> Glad to hear that!
> 
> > and I plan to submit a PR to fix the problem first, and then add UT. Is that okay?
> 
> Sure, you can. I think, for a single pr, sometimes, small is better :D

issue: https://github.com/apache/dolphinscheduler/issues/15114
pr: https://github.com/apache/dolphinscheduler/pull/15116

**Pull request merged**